glossary: Add more abbreviations (#19213)
pingcap/docs: glossary.md

# Glossary

This glossary provides definitions for key terms related to the TiDB platform.

Other available glossaries:

> - [DM Glossary](/dm/dm-glossary.md)

Reviewer: ```suggestion
- [TiDB Data Migration Glossary](/dm/dm-glossary.md)
```

## B

### Backup & Restore (BR)

BR is the backup and restore tool for TiDB. For more information, see [BR Overview](/br/backup-and-restore-overview.md).

`br` is the [br command line tool](/br/use-br-command-line-tool.md) used for backups or restores in TiDB.

### Baseline Capturing

Baseline Capturing captures queries that meet capturing conditions and create bindings for them. It is used for [preventing regression of execution plans during an upgrade](/sql-plan-management.md#prevent-regression-of-execution-plans-during-an-upgrade).

### Batch Create Table

> Batch Create Table is a feature introduced in TiDB v6.0.0. This feature is enabled default. When restoring data with a large number of tables (nearly 50000) using BR (Backup & Restore), the feature can greatly speed up the restore process by creating tables in batches. For details, see [Batch Create Table](/br/br-batch-create-table.md).

Reviewer 2: ```suggestion
Batch Create Table is a feature introduced in TiDB v6.0.0. This feature is enabled by default. When restoring data with a large number of tables (nearly 50000) using BR (Backup & Restore), the feature can greatly speed up the restore process by creating tables in batches. For details, see [Batch Create Table](/br/br-batch-create-table.md).
```

### Data Migration (DM)

Data Migration (DM) is a tool for migrating data from MySQL-compatible databases into TiDB. DM reads data from a MySQL-compatible database instance and applies it to a TiDB target instance. For more information, see [DM Overview](/dm/dm-overview.md).

### Data Modification Language (DML)

Data Modification Language (DML) is a part of the SQL standard that deals with inserting, updating, and dropping rows in tables.

### Development Milestone Release (DMR)

Development Milestone Releases (DMR) are TiDB releases that introduce the latest features but do not offer long-term support. For more information, see [TiDB Versioning](/releases/versioning.md).

### Disaster Recovery (DR)

Disaster Recovery (DR) includes solutions that can be used to recover data and services from a disaster in the future. TiDB offers various Disaster Recovery solutions such as backups and replication to standby clusters. For more information, see [Overview of TiDB Disaster Recovery Solutions](/dr-solution-introduction.md).

### Distributed eXecution Framework (DXF)

> Distributed eXecution Framework (DXF) is the framework used by TiDB to distribute tasks across a TiDB cluster. DXF is designed to efficiently use cluster resources to execute tasks (such as index creation or data import) while controlling the resource usage and impact on core business transactions. For more information, see [DXF Introduction](/tidb-distributed-execution-framework.md).

Reviewer: ```suggestion
Distributed eXecution Framework (DXF) is the framework used by TiDB to centrally schedule and execute certain tasks (such as creating indexes or importing data) in a distributed manner. DXF is designed to efficiently use cluster resources while controlling resource usage and reducing the impact on core business transactions. For more information, see [DXF Introduction](/tidb-distributed-execution-framework.md).
```

Author: The tasks are centrally scheduled, but not centrally executed. Maybe something like "centrally schedule ... and execute in distributed manner"?

Reviewer: Done. PTAL again, thanks.

Introduced in TiDB 5.3.0, Continuous Profiling is a way to observe resource overhead at the system call level. With the support of Continuous Profiling, TiDB provides performance insight as clear as directly looking into the database source code, and helps R&D and operation and maintenance personnel to locate the root cause of performance problems using a flame graph. For details, see [TiDB Dashboard Instance Profiling - Continuous Profiling](/dashboard/continuous-profiling.md).

### CTE

A Common Table Expression (CTE) is part of the SQL standard and uses [`WITH`](/sql-statements/sql-statement-with.md) statements.

## D

### DDL

> Data Definition Language (DDL) is the part of the SQL standard that deals with creating, modifying and deleting tables and other objects.

Reviewer 3: ```suggestion
Data Definition Language (DDL) is the part of the SQL standard that deals with creating, modifying and deleting tables, indexes, columns and other objects.
```

### DM

Data Migration (DM) is a tool for migrating data from MySQL-compatible databases into TiDB. It reads data from an instance of MySQL-compatible database and applies it to a TiDB target instance. For more information, see [DM Overview](/dm/dm-overview.md).

### DML

Data Modification Language (DML) statements enables you to with insert, update, and delete rows in tables.

### DMR

Development Milestone Release (DMR) is a TiDB version that introduces the latest features but does not offer long-term support. For more information, see [TiDB Versioning](/releases/versioning.md).

### DR

Disaster Recovery (DR) includes solutions that can be used to recover data from a disaster in the future. These solutions typically involve backups and standby clusters. For more information, see [Overview of TiDB Disaster Recovery Solutions](dr-solution-introduction).

### DXF

> Distributed eXecution Framework (DXF) is the framework used by TiDB for accelerating index creation and data import by distributing tasks over all available resources. For more information, see [DXF Introduction](/tidb-distributed-execution-framework.md).

Reviewer 4: Some suggestions to make it clear DXF scope is broader than just index creation and data import tasks, e.g. these are examples, not the only tasks that may be distributed:

Distributed eXecution Framework (DXF) is the framework used by TiDB to distribute tasks across the TiDB cluster. DXF is designed to efficiently use the cluster resources to execute tasks (like index creation or data import) while controlling the resource usage and impact on core business transactions. For more information, see [DXF Introduction](/tidb-distributed-execution-framework.md).

### Dynamic Pruning

Dynamic pruning mode is one of the modes that TiDB accesses partitioned tables. In dynamic pruning mode, each operator supports direct access to multiple partitions. Therefore, TiDB no longer uses Union. Omitting the Union operation can improve the execution efficiency and avoid the problem of Union concurrent execution.

## E

### EC2

Elastic Compute Cloud (EC2) is an AWS service that provides compute resources. This can be used with TiUP to run a TiDB Cluster.

## G

### GA

> General Available (GA) is the first non-beta version of a software product.

Author: I think keeping this as "first non-beta version" is best. I think "first production quality release" might add a quality assurance that might not hold true for all productions (including non-PingCAP ones)

> ### EC2

Reviewer 4: Do we want to define EC2 in our glossary? Practically speaking, will people be looking to us to define EC2 for them in this doc? I worry that this would expand to defining a bunch of other third party terms if we go down this path.

> ### GA

Reviewer 4: General Availability (GA)

> If a feature is General Available (GA), it indicates it is fully tested and can be used in production environments. Note that even if a feature is GA in a [DMR](#dmr) version, it is recommended to use the feature in production environments in a later [LTS](#lts) version.

Reviewer 4: Suggested edit, we should use General Availability (to indicate the feature release stage) or Generally Available (to indicate the status of the feature):

General Availability (GA) of a feature is when it is, fully tested and is Generally Available for use in production environments. TiDB features may be released as Generally Available in both [DMR](#dmr) and [LTS](#lts) releases. However, as TiDB does not provide patch releases based on DMR it is generally recommended to use the LTS product release for production use.

> ### GC

Reviewer 4: Garbage Collection (GC)

Garbage Collection (GC) is a process that clears obsolete data to free up resources. For information on TiKV GC process, see [Garbage Collection overview](/garbage-collection-overview.md).

> ### GTID

Reviewer 4: Global Transaction Identifiers (GTIDs)

Global Transactions ID's (GTIDs) are used by recent MySQL versions to indicate what transactions have been replicated and which have not. This information can be used by DM.

## H

### HTAP

> Hybrid Transactional Analytical Process (HTAP) is a database feature that allows both OLTP and OLAP workloads on the same database. For TiDB the HTAP feature is provided by using both TiKV for row storage and TiFlash for columnar storeage. See [the definition of HTAP on the Gartner website](https://www.gartner.com/en/information-technology/glossary/htap-enabling-memory-computing-technologies) for more information.

Reviewer 3: ```suggestion
Hybrid Transactional Analytical Process (HTAP) is a database feature that allows both OLTP and OLAP workloads on the same database. For TiDB the HTAP feature is provided by using both TiKV for row storage and TiFlash for columnar storage. See [the definition of HTAP on the Gartner website](https://www.gartner.com/en/information-technology/glossary/htap-enabling-memory-computing-technologies) for more information.
```

> ### HTAP

Reviewer 4: Hybrid Transactional and Analytical Processing (HTAP)

## I

> ### IMDS

Reviewer 4: Not sure why we want to include this third party abbreviation here

Author: This is used in `br/backup-and-restore-storages.md` and in the [TiDB v8.4.0 release notes](https://github.com/pingcap/docs/pull/18829).

### Disaster Recovery (DR)

Disaster Recovery (DR) includes solutions that can be used to recover data and services from a disaster in the future. TiDB offers various Disaster Recovery solutions such as backups and replication to standby clusters. For more information, see [Overview of TiDB Disaster Recovery Solutions](/dr-solution-introduction.md).

### Distributed eXecution Framework (DXF)

Distributed eXecution Framework (DXF) is the framework used by TiDB to distribute tasks across a TiDB cluster. DXF is designed to efficiently use cluster resources to execute tasks (such as index creation or data import) while controlling the resource usage and impact on core business transactions. For more information, see [DXF Introduction](/tidb-distributed-execution-framework.md).

### Dynamic Pruning

Dynamic pruning mode is one of the modes that TiDB accesses partitioned tables. In dynamic pruning mode, each operator supports direct access to multiple partitions. Therefore, TiDB no longer uses Union. Omitting the Union operation can improve the execution efficiency and avoid the problem of Union concurrent execution.

## G

### Garbage Collection (GC)

> Garbage Collection (GC) is a process that clears obsolete data to free up resources. For information on TiKV GC process, see [Garbage Collection overview](/garbage-collection-overview.md).

Reviewer 2: ```suggestion
Garbage Collection (GC) is a process that clears obsolete data to free up resources. For information on TiKV GC process, see [Garbage Collection Overview](/garbage-collection-overview.md).
```

The in-memory pessimistic lock is a new feature introduced in TiDB v6.0.0. When this feature is enabled, pessimistic locks are usually stored in the memory of the Region leader only, and are not persisted to disk or replicated through Raft to other replicas. This feature can greatly reduce the overhead of acquiring pessimistic locks and improve the throughput of pessimistic transactions.

## K

> ### KMS

Reviewer 4: Key Management Service (KMS)

> Key Management Service (KMS) enables the storage and retrieval of secret keys in a secure way. Examples include AWS KMS, GCP KMS, and HashiCorp Vault. Various TiDB components can use KMS to manage keys for storage encryption and related services.

Reviewer: ```suggestion
Key Management Service (KMS) enables the storage and retrieval of secret keys in a secure way. Examples include AWS KMS, Google Cloud KMS, and HashiCorp Vault. Various TiDB components can use KMS to manage keys for storage encryption and related services.
```

> ### KV

Reviewer 4: Key-Value (KV)

Key-Value (KV) is a way of storing information by associating values with unique keys, allowing quick data retrieval. TiDB uses TiKV to map tables and indexes into key-value pairs, enabling efficient data storage and access across the database.

## L

> ### LDAP

Reviewer 4: Lightweight Directory Access Protocol (LDAP)

Lightweight Directory Access Protocol (LDAP) is a standardized way of accessing a directory with information. It is commonly used for account and user data management. TiDB supports LDAP via [LDAP authentication plugins](/security-compatibility-with-mysql.md#authentication-plugin-status).

> ### LTS

Reviewer 4: Long Term Support (LTS)

Long Term Support (LTS) refers to software versions that are extensively tested and maintained for extended periods. For more information, see [TiDB Versioning](/releases/versioning.md).

> ### leader/follower/learner

Reviewer 4: Should be capitalized:
Leader/Follower/Learner

## O

> ### OLAP

Reviewer 4: Online Analytical Processing (OLAP)

Online Analytical Processing (OLAP) refers to database workloads focused on analytical tasks, such as data reporting and complex queries. OLAP is characterized by read-heavy queries that process large volumes of data across many rows.

### Old value

The "original value" in the incremental change log output by TiCDC. You can specify whether the incremental change log output by TiCDC contains the "original value".

> ### OLTP

Reviewer 4: Online Transaction Processing (OLTP)

Online Transaction Processing (OLTP) refers to database workloads focused on transactional tasks, such as selecting, inserting, updating, and deleting small sets of records.

> ## OOM

Reviewer 4: Out of Memory (OOM)

[Partitioning](/partitioned-table.md) refers to physically dividing a table into smaller table partitions, which can be done by partition methods such as RANGE, LIST, HASH, and KEY partitioning.

> ### PD

Reviewer 4: Placement Driver (PD) Server

> Placement Driver (PD) is an important component of the [TiDB Architecture](/tidb-architecture.md#placement-driver-pd-server) that is responsible to store metadata and run the [TSO](/tso.md) that hands out timestamps that are used for transactions. It also orchestrates the data placement on TiKV and runs the [TiDB Dashboard](/dashboard/dashboard-overview.md).

Author: Not sure, it might. Maybe there are other things that I forgot about ("PD HTTP API"?). Does it have to be an exhaustive list? I think the current list has the most important tasks.

> Placement Driver (PD) is an important component of the [TiDB Architecture](/tidb-architecture.md#placement-driver-pd-server) that is responsible to store metadata and run the [TSO](/tso.md) that hands out timestamps that are used for transactions. It also orchestrates the data placement on TiKV and runs the [TiDB Dashboard](/dashboard/dashboard-overview.md).

Author: The PD section of tidb-architecture.md where this links to also doesn't mention TiFlash.

> Placement Driver (PD) is an important component of the [TiDB Architecture](/tidb-architecture.md#placement-driver-pd-server) that is responsible to store metadata and run the [TSO](/tso.md) that hands out timestamps that are used for transactions. It also orchestrates the data placement on TiKV and runs the [TiDB Dashboard](/dashboard/dashboard-overview.md).

Reviewer 3: OK, thanks for checking!

### pending/down

"Pending" and "down" are two special states of a peer. Pending indicates that the Raft log of followers or learners is vastly different from that of leader. Followers in pending cannot be elected as leader. "Down" refers to a state that a peer ceases to respond to leader for a long time, which usually means the corresponding node is down or isolated from the network.

> ### PITR

Reviewer 4: Point in Time Recovery (PITR)

Raft Engine is an embedded persistent storage engine with a log-structured design. It is built for TiKV to store multi-Raft logs. Since v5.4, TiDB supports using Raft Engine as the log storage engine. For details, see [Raft Engine](/tikv-configuration-file.md#raft-engine).

> ### Region split
> 
> Regions are generated as data writes increase. The process of splitting is called Region split.

Reviewer 2: ```suggestion
### Region Split

A region in a TiKV cluster is not divided at the beginning, but is gradually split as data is written to it. The process is called Region split.
```

Author: Not sure if it is clear enough to a reader that this is a TiKV data region and not an AWS region etc

> A region in a TiKV cluster is not divided at the beginning, but is gradually split as data is written to it. The process is called Region split.

Reviewer: ```suggestion
A region in a TiKV cluster is not divided at the beginning but is gradually split as data is written to it. The process is called Region split.
```

The mechanism of Region split is to use one initial Region to cover the entire key space, and generate new Regions through splitting existing ones every time the size of the Region or the number of keys has reached a threshold.

> ### Region/peer/Raft group

Reviewer 2: ```suggestion
### Region/Peer/Raft Group
```

The mechanism of Region split is to use one initial Region to cover the entire key space, and generate new Regions through splitting existing ones every time the size of the Region or the number of keys has reached a threshold.

### Restore

Restore is the reverse of the backup operation. It is the process of bringing back the system to an earlier state by retrieving data from a prepared backup.

> ### RPC

Reviewer 4: Remote Procedure Call (RPC)

Remote Procedure Call (RPC) is a communication way between software components. In a TiDB cluster, the gRPC standard is used for communication between different components such as TiDB, TiKV, and TiFlash.

> ### RU

Reviewer 4: Request Unit (RU)

- `hot-region-scheduler`: Balances the distribution of hot Regions
- `evict-leader-{store-id}`: Evicts all leaders of a node (often used for rolling upgrades)

> ### SST

Reviewer 4: Static Sorted Table / Sorted String Table (SST)

> Static Sorted Table, Sorted String Table, or Sorted Sequence Table (SST) is a file storage format used in RocksDB.

Reviewer 4: I don't believe we use the term Sorted Sequence Table in our product. Added a pointer to the rocksdb overview in suggested edit below:

Static Sorted Table or Sorted String Table is a file storage format used in RocksDB (a component used by the [TiKV Storage Engine](rocksdb-overview.md)).

### Store

> A store refers to the storage node in the TiKV cluster (an instance of `tikv-server`). Each store has a corresponding TiKV instance.

Reviewer 4: As we adding the PD component to this glossary should we also add the TiDB Server, TiFlash Server, and TiKV Server components as well for completeness?

Suggested additions:

### TiDB Server

The [TiDB server](https://docs.pingcap.com/tidb/stable/tidb-computing) is a stateless SQL layer that exposes the connection endpoint of the MySQL protocol to the outside. The TiDB server receives SQL requests, performs SQL parsing and optimization, and ultimately generates a distributed execution plan.

### TiFlash Server

The [TiFlash server](https://docs.pingcap.com/tidb/stable/tiflash-overview) is a special type of storage server. Unlike ordinary TiKV nodes, TiFlash stores data by column, mainly designed to accelerate analytical processing.

### TiKV Server

The [TiKV server](https://docs.pingcap.com/tidb/stable/tidb-storage) is responsible for storing data. TiKV is a distributed transactional key-value storage engine.

Top SQL helps locate SQL queries that contribute to a high load of a TiDB or TiKV node in a specified time range. For details, see [Top SQL user document](/dashboard/top-sql.md).

> ### TPS

Reviewer 4: Transactions Per Second (TPS)

Transactions Per Second (TPS) is the number of transactions a database processes per second, serving as a key metric for measuring database performance and throughput.

> ### TSO

Reviewer 4: Timestamp Oracle (TSO)

Because TiKV is a distributed storage system, it requires a global timing service, Timestamp Oracle (TSO), to assign a monotonically increasing timestamp. In TiKV, such a feature is provided by PD, and in Google [Spanner](http://static.googleusercontent.com/media/research.google.com/en//archive/spanner-osdi2012.pdf), this feature is provided by multiple atomic clocks and GPS. For details, see [TSO](/tso.md).

## U

> ### URI

Reviewer 4: Uniform Resource Identifier (URI)

> Uniform Resource Identifier (URI) is a standardized format for identifying a resource. For more information, see [Uniform Resource Identifier](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) on Wikipedia.

Reviewer 4: Why do we want to include this definition in the glossary?

Author: This is because it is mentioned in our docs quite often. I would assume URI/URL could be considered common knowledge and left out. I'm fine with it either way.

```
$ git grep -cw URI | sed "s/:/\t/" | sort -k2 -r | head
ticdc/ticdc-sink-to-mysql.md	7
tidb-cloud/config-s3-and-gcs-access.md	6
tidb-cloud/changefeed-sink-to-cloud-storage.md	6
ticdc/ticdc-sink-to-kafka.md	6
dumpling-overview.md	6
br/br-pitr-manual.md	5
tiproxy/tiproxy-api.md	4
tidb-cloud/tidb-cloud-auditing.md	4
tidb-cloud/migrate-sql-shards.md	4
tidb-cloud/migrate-from-op-tidb.md	4
```

Author: @qiancai @benmeadowcroft should we keep it or not? It is something well known and not TiDB specific. But on the other hand, it is often mentioned in the docs.

> ### UUID

Reviewer 4: Universally Unique Identifier (UUID)
